### PR TITLE
adding a pull request template

### PR DESCRIPTION
**Pull request by rheacangeo**
## Description
 Add the PR template and point the CONTRIBUTING document to our paper dropbox location where we maintain a list of code style guidelines in the form of a code review checklist.

### CODE REVIEW

**Comment by rheacangeo** [comment]
launch jenkins

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+Describe the content of the PR and links to related issues, bugs of features.
+
+## Requirements
+
+Before submitting this PR, please make sure:
+
+- [ ] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
+- [ ] The code builds cleanly without new errors or warnings
+- [ ] The code passes all the existing tests
```
Comment (mcgibbon) [inline .github/pull_request_template.md:11]: These shouldn't be needed as part of the checklist, because it is already enforced by Github and our CI system. We already can't merge PRs which don't satisfy these.

I would suggest taking this bullet in particular from the code review checklist, as it's probably the number one easiest and most impactful change authors can include/add to their other changes (of course giving exceptions where we don't have the necessary info, as appropriate).

You could if you prefer combine this bullet with the next one (since they are both about having documentation).
```suggestion
- [ ] Docstrings and type hints are added to new and updated routines, as appropriate
```
```suggestion
- [ ] All relevant documentation has been updated or added. Docstrings and type hints are added to new and updated routines, as appropriate
```

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+## Description
+
+Describe the content of the PR and links to related issues, bugs of features.
```
Comment (mcgibbon) [inline .github/pull_request_template.md:3]: I would suggest expanding this out a bit more, including some (but not all) of the content from the [fv3net template.](https://github.com/VulcanClimateModeling/fv3net/blob/master/.github/pull_request_template.md).

A good set of headings might be "Code changes" with a bullet, and "requirement changes" with a bullet. We don't have much public API or multiple packages with inter-dependencies the way fv3net does, so the stress on internal vs public changes isn't as important. This might change later and we could re-visit it.

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+## Requirements
+
+Before submitting this PR, please make sure:
+
+- [ ] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
+- [ ] The code builds cleanly without new errors or warnings
+- [ ] The code passes all the existing tests
+- [ ] All relevant documentation has been updated or added
```
Comment (mcgibbon) [inline .github/pull_request_template.md:12]: Since this hasn't been a common practice up to now, I think it's worth explicitly pasting in from the full checklist. Of course there can still be exceptions but this is one that's likely to be left out simply because the author didn't think of it.
```suggestion
- [ ] All relevant documentation has been updated or added
- [ ] Unit tests are added or updated for non-stencil code changes
```

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+
+Additionally, if this PR contains code authored by new contributors:
+
+- [ ] The names of all the new contributors have been added to an updated
+version of the CONTRIBUTORS file included in the PR
```
Comment (mcgibbon) [inline .github/pull_request_template.md:18]: Removed some unneeded words, I found it hard to parse before.

```suggestion
- [ ] The names of all the new contributors have been added to CONTRIBUTORS.md
```

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
+## Description
+
+Describe the content of the PR and links to related issues, bugs of features.
+
+## Requirements
+
+Before submitting this PR, please make sure:
+
+- [ ] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
```
Comment (mcgibbon) [inline .github/pull_request_template.md:9]: Great! In practice most people are unlikely to read the full checklist each time they write/review a PR, at least after the first few times. Because of this I've suggested below adding two points from that checklist explicitly to this one (adding the whole checklist here would be overkill).

**Review by mcgibbon — COMMENTED** [review]
Great start! I would suggest expanding the description template a little bit, and some modifications to the requirements bullets. Happy to discuss the recommendations if you disagree.

File: .github/pull_request_template.md (hunk 2 above)
Comment (rheacangeo) [reply .github/pull_request_template.md]: Thanks for sharing that, helpful!

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+
+## Purpose
+
+Describe the purpose of this PR and the motivation if relevant to understanding. Include links to related issues, bugs or features. Include the sections below that apply.
+
+## Code changes:
+Provide a list of relevant code changes and the side effects they have on other code.
+
```
Comment (mcgibbon) [inline .github/pull_request_template.md:8]:
```suggestion
## Code changes:

- Provide a list of relevant code changes and the side effects they have on other code.

```

File: .github/pull_request_template.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+## Requirements changes:
+Provide a list of any changes made to requirements or linting, e.g. changes to files requirements*.txt, constraints.txt, setup.py, pyproject.toml, pre-commit-config.yaml as well as a reason if not included in the Purpose section (e.g. incompatibility, updates, etc)
+
```
Comment (mcgibbon) [inline .github/pull_request_template.md:11]: Changes to linting requirements will already be reflected in code changes (because the code will need to be updated) and infrastructure changes (we are changing the testing infrastructure). If you like you could change the section heading to "Dependency changes".

```suggestion
## Requirements changes:

- Provide a list of any changes made to requirements, e.g. changes to files requirements*.txt, constraints.txt, setup.py, pyproject.toml, pre-commit-config.yaml and a reason if not included in the Purpose section (e.g. incompatibility, updates, etc)

```

File: .github/pull_request_template.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+## Infrastructure changes:
+Provide a list of changes that impact the infrastructure around running the code -- that is, changes to Makefiles, docker files, git submodules, or .jenkins (testing infrastructure changes). If Jenkins plans are also being manually changed, indicate that as well.
+
```
Comment (mcgibbon) [inline .github/pull_request_template.md:14]:
```suggestion
## Infrastructure changes:

- Provide a list of changes that impact the infrastructure around running the code -- that is, changes to Makefiles, docker files, git submodules, or .jenkins (testing infrastructure changes). If Jenkins plans are also being manually changed, indicate that as well.

```

File: .github/pull_request_template.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+## Purpose
+
+Describe the purpose of this PR and the motivation if relevant to understanding. Include links to related issues, bugs or features. Include the sections below that apply.
```
Comment (mcgibbon) [inline .github/pull_request_template.md:4]:
```suggestion
Describe the purpose of this PR and the motivation if relevant to understanding. Include links to related issues, bugs or features.

Remove the sections below which do not apply.
```

**Review by mcgibbon — APPROVED** [review]
Approved with small optional change suggestions.

It would be helpful to start everyone off with bulleted lists, so it's clear we expect a bulleted list syntax and avoid a mixture of bullet- and paragraph-formatted PRs.

File: .github/pull_request_template.md (hunk 7 above)
Comment (rheacangeo) [reply .github/pull_request_template.md:12]: I meant more changes to the linting tools -- e.g. the configuration of the linting, but we don't have to say linting.
